lib/buildpacks_registry_api/cache_client.ex

```diff
@@ -0,0 +1,349 @@
+        name: "ruby",
+        namespace: "heroku",
+        stacks: ["heroku-18", "heroku-20"],
+        version: "0.1.3"
+      },
+      versions: [
+        %{
+          _link: "https://registry.buildpacks.io//api/v1/buildpacks/heroku/ruby/0.1.3",
+          version: "0.1.3"
+        },
+        %{
+          _link: "https://registry.buildpacks.io//api/v1/buildpacks/heroku/ruby/0.1.2",
+          version: "0.1.2"
+        },
+        %{
+          _link: "https://registry.buildpacks.io//api/v1/buildpacks/heroku/ruby/0.1.1",
+          version: "0.1.1"
+        },
+        %{
+          _link: "https://registry.buildpacks.io//api/v1/buildpacks/heroku/ruby/0.1.0",
+          version: "0.1.0"
+        }
+      ]
+    }
+  """
+  @spec buildpack_version_list(binary(), binary()) :: map()
+  def buildpack_version_list(namespace, name) do
+    get("/buildpacks/#{namespace}/#{name}", %{})
+  end
+
+  @doc """
+  Returns detailed information for the buildpack version.any()
+
+  ## Examples
+    iex> BuildpacksRegistryApi.Client.buildpack_version_info("heroku", "ruby", "0.1.3")
+    %{
+      addr:
+        "public.ecr.aws/heroku-buildpacks/heroku-ruby-buildpack@sha256:3eeb3773cdbd29d4fb0d578f7781fe8c525de73593480e2740b7143262e5bef5",
+      created_at: "2021-10-27T21:00:43.307Z",
+      description: "",
+      homepage: "",
+      id: "f7c0dcc3-55ab-46b9-9a89-e35ff60aa7f8",
+      licenses: nil,
+      name: "ruby",
+      namespace: "heroku",
+      stacks: ["heroku-18", "heroku-20"],
+      updated_at: "2022-03-17T21:20:26.981Z",
+      version: "0.1.3",
+      version_major: "0",
+      version_minor: "1",
+      version_patch: "3",
+      yanked: false
+    }
+  """
+  @spec buildpack_version_info(binary(), binary(), binary()) :: map()
+  def buildpack_version_info(namespace, name, version) do
+    get("/buildpacks/#{namespace}/#{name}/#{version}", %{})
+  end
+
+  # Server callbacks
+  @impl true
+  def init(state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_cast(:clear, state) do
+    Logger.debug("Clearing cache")
+    {:noreply, %{state | cache: %{}}}
+  end
+
+  @impl true
+  def handle_cast({:set_cache_timeout, timeout_in_ms}, state) do
+    Logger.debug("Setting cache timeout to #{timeout_in_ms}")
+    {:noreply, %{state | cache_timeout: timeout_in_ms}}
+  end
+
+  @impl true
+  def handle_call(:config, _from, state) do
+    {:reply,
+     %{
+       cache_timeout: state.cache_timeout
+     }, state}
+  end
+
+  @impl true
+  def handle_call({:return_cache_or_get, path, params}, _from, %State{} = state) do
+    url = BuildpacksRegistryApi.Client.process_url(path, params)
+    now = Time.utc_now()
+
+    case state.cache |> Map.get(url) do
+      nil ->
+        Logger.debug("#{url} cache miss")
+        response = Client.get(path, params) |> Jason.decode!(keys: :atoms)
+        new_cache = Map.merge(state.cache, %{url => %{response: response, fetched_at: now}})
+        new_state = %State{state | cache: new_cache}
+
```

Comment:
Could be extracted to a function for reuse when the cache has expired, ~also this appears to assume that the get will always pass, do we want to cache errors?~ (found that `get` raises)

Reply:
Fetch and cache has been moved into a private function.